### PR TITLE
Support inline rendering of CUSTOM_URL_SCHEMES

### LINKED ISSUE #8478: Display of FTP links in markdown
- Gitea version (or commit ref): 1.9.3
- Git version: 2.1.9
- Operating system: Cent OS 7
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [ ] MySQL
  - [X] MariaDB
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ X] Yes (provide example URL) https://try.gitea.io/bkmgit/TestMarkdown/src/branch/master/README.md
  - [ ] No
  - [ ] Not relevant
- Log gist:

## Description
Markdown does not render ftp links correctly
Similar issue appears [here](https://github.com/gollum/gollum/issues/759)
...

## Screenshots

FTP_links

### LINKED ISSUE #8490: Proposal: list of URL schemes to support in markdown rendering
Currently we're allowing only http and https schemes (protocols) in links rendered markdown. There are users that want others, like ftp (e.g. #8478)

Currently, this renders nothing:
```
[My link](ftp://www.mysite.com)
```

There are a number of [other schemes we could support](https://en.wikipedia.org/wiki/List_of_URI_schemes), and some of them could be very useful, like the aforementioned ftp, VS Code extensions, cvs, ed2k, torrent, git, ssh, magnet, maps, sftp, skype, svn, etc.

I'm not sure which ones could pose a security problem, so I propose adding an `app.ini` settings with the list of desired schemes. The obvious default is `http, https`, but perhaps we want to add `ftp` and others, so let's hear out your ideas.

### PR DESCRIPTION
**Pull request by guillep2k**
Adds support for inline rendering of URL schemes specified in CUSTOM_URL_SCHEMES

```
http [google](http://www.google.com/)

https [google](https://www.google.com/)

ftp [google](ftp://www.google.com/)

http://www.google.com/

https://www.google.com/

ftp://www.google.com/

gogo://www.google.com/

magnet:?xt=urn:btih:5dee65101db281ac9c46344cd6b175cdcad53426&dn=download

magnet [google](magnet:?xt=urn:btih:5dee65101db281ac9c46344cd6b175cdcad53426&dn=download)
```

image

For `app.ini`:

```
[markdown]
CUSTOM_URL_SCHEMES = ftp, magnet
```

Fixes #8490 
Fixes #8478 
Related to #8495

### CODE REVIEW

**Comment by silverwind** [comment]
Please add new config to [app.ini.sample](https://github.com/go-gitea/gitea/blob/master/custom/conf/app.ini.sample).

File: modules/markup/html.go [hunk 1]
```diff
@@ -92,6 +92,32 @@ func getIssueFullPattern() *regexp.Regexp {
 	return issueFullPattern
 }
 
+// CustomLinkURLSchemes allows for additional schemes to be detected when parsing links within text
+func CustomLinkURLSchemes(schemes []string) {
+	schemes = append(schemes, "http", "https")
```
Comment (silverwind) [inline modules/markup/html.go:97]: Do we need to add `mailto` here? It looks to currently render correctly.

**Comment by guillep2k** [comment]
> 
> 
> Please add new config to [app.ini.sample](https://github.com/go-gitea/gitea/blob/master/custom/conf/app.ini.sample).

There's no new config. I've only added support for *inline* detection. Old explicit links already worked:
```
[name](scheme:link)  <--- this already worked
scheme:link          <--- this is now supported
```

File: modules/markup/html.go (hunk 1 above)
Comment (guillep2k) [reply modules/markup/html.go:97]: e-mail addresses are parsed separately. The `mailto:` URL scheme however must be added explicitly if wanted:

```
Hey, user anon@anon.com !                 <--- this doesn't need any settings
Hey, user mailto:anon@anon.com !          <--- this requires adding mailto to the list of custom schemes, and will need my PR to work
Hey, user [name](mailto:anon@anon.com)    <--- this too requires adding mailto to the list of custom schemes (not related to my PR)
```

**Review by zeripath — APPROVED** [review]

**Review by 6543 — APPROVED** [review]

**Pull request merged**